freva-data-portal-worker/src/data_portal_worker/load_data.py

```diff
@@ -0,0 +1,324 @@
+        data = encode_chunk(
+            get_data_chunk(
+                encode_zarr_variable(dset.variables[variable], name=variable).data,
+                chunk,
+                out_shape=arr_meta["chunks"],
+            ).tobytes(),
+            filters=arr_meta["filters"],
+            compressor=arr_meta["compressor"],
+        )
+        self.cache.setex(f"{key}-{variable}-{chunk}", 360, data)
+
+    def load_object(self, key: str) -> LoadDict:
+        """Load a cached dataset.
+
+        Parameters
+        ----------
+        key: str, The cache key.
+
+        Returns
+        -------
+        The data that was stored under that key.
+
+        Raises
+        ------
+        RuntimeError: If the cache key exists but the data could not be loaded,
+                      which means that there is a load status != 0
+        KeyError: If the key doesn't exist in the cache (anymore).
+        """
+        data_cache = self.cache.get(key)
+        if data_cache is None:
+            raise KeyError(f"{key} uuid does not exist (anymore).")
+        return cast(LoadDict, cloudpickle.loads(data_cache))
+
+
+class ProcessQueue(DataLoadFactory):
+    """Class that can load datasets on different object stores."""
+
+    def __init__(self, dev_mode: bool = False) -> None:
+        super().__init__()
+        self.client = get_dask_client(dev_mode=dev_mode)
+        self.dev_mode = dev_mode
+
+    def run_for_ever(self, channel: str) -> None:
+        """Start the listner deamon."""
+        data_logger.info("Starting data-loading deamon")
+        pubsub = self.cache.pubsub()
+        pubsub.subscribe(channel)
+        data_logger.info("Broker will listen for messages now")
+        while True:
+            message = pubsub.get_message()
+            if message and message["type"] == "message":
+                try:
+                    self.redis_callback(message["data"])
+                except KeyboardInterrupt:
+                    raise
+                except Exception as error:
+                    data_logger.exception(error)
+            else:
+                time.sleep(0.1)
+
+    def redis_callback(
+        self,
+        body: bytes,
+    ) -> None:
+        """Callback method to recieve rabbit mq messages."""
+        try:
+            message = json.loads(body)
+        except json.JSONDecodeError:
+            data_logger.warning("could not decode message")
+            return
+        if "uri" in message:
+            self.spawn(message["uri"]["path"], message["uri"]["uuid"])
+        elif "chunk" in message:
+            self.get_zarr_chunk(
+                message["chunk"]["uuid"],
+                message["chunk"]["chunk"],
+                message["chunk"]["variable"],
+            )
+        elif "shutdown" in message:
+            if message["shutdown"] is True and self.dev_mode is True:
+                raise KeyboardInterrupt("Shutdown client")
+
+    def spawn(self, inp_obj: str, uuid5: str) -> str:
```

Comment:
I think we want first add an empty entry and then start the processing. This might prevent race conditions. Processing might take long. And if a user spawns many requests at the same time the same jobs might be spawned because the entries are only added once the data has been loaded. But I think we want to add something to the cache fast, even if it's empty data.